src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/DocumentGeneratedOutputTracker.cs

```diff
@@ -62,22 +65,29 @@ public DocumentGeneratedOutputTracker Fork()
         private async Task<RazorCodeDocument> GetGeneratedOutputInitializationTaskCore(ProjectSnapshot project, DocumentSnapshot document)
         {
             var tagHelpers = await project.GetTagHelpersAsync().ConfigureAwait(false);
+            var imports = await GetImportsAsync(project, document);
+
             if (_older != null && _older.IsResultAvailable)
             {
-                var difference = new HashSet<TagHelperDescriptor>(TagHelperDescriptorComparer.Default);
-                difference.UnionWith(_older._tagHelpers);
-                difference.SymmetricExceptWith(tagHelpers);
+                var tagHelperDifference = new HashSet<TagHelperDescriptor>(TagHelperDescriptorComparer.Default);
+                tagHelperDifference.UnionWith(_older._tagHelpers);
+                tagHelperDifference.SymmetricExceptWith(tagHelpers);
+
+                var importDifference = new HashSet<ImportItem>();
+                importDifference.UnionWith(_older._imports);
+                importDifference.SymmetricExceptWith(imports);
 
-                if (difference.Count == 0)
+                if (tagHelperDifference.Count == 0 && importDifference.Count == 0)
```

Comment:
This needs to also include checks for document version. That will be done in a follow up.

Reply:
Cool beans. 

Yeah, you pushed this much further in this PR than I expected. It good that you are hyped about taking all of this on, but we should try to iterate faster.